### PR TITLE
Improve post and pre migration checks for pallet-membership

### PR DESCRIPTION
**Pull request by thiolliere**
some of the assertion weren't checking much, this is an improved version

### CODE REVIEW

File: frame/membership/src/migrations/v4.rs [hunk 1]
```diff
@@ -90,23 +89,17 @@ pub fn pre_migrate<P: GetStorageVersion, N: AsRef<str>>(old_pallet_name: N, new_
...
+	let storage_version_key = twox_128(STORAGE_VERSION_STORAGE_KEY_POSTFIX);
+
+	let mut new_pallet_prefix_iter = frame_support::storage::KeyPrefixIterator::new(
+		new_pallet_prefix.to_vec(),
+		new_pallet_prefix.to_vec(),
+		|key| Ok(key.to_vec()),
 	);
+
+	// Ensure nothing except the storage_version_key is stored in the new prefix.
+	assert!(new_pallet_prefix_iter.all(|key| key == storage_version_key));
```
Comment (thiolliere) [inline frame/membership/src/migrations/v4.rs:101]: in the past we only ensured that the first key was sotrage_version_key, now we ensure it is the only one.

File: frame/membership/src/migrations/v4.rs [hunk 2]
```diff
@@ -119,26 +112,24 @@ pub fn post_migrate<P: GetStorageVersion, N: AsRef<str>>(old_pallet_name: N, new
...
-		// Assert that nothing remains at the old prefix
-		assert!(storage::next_key(&old_pallet_prefix)
-			.map_or(true, |next_key| !next_key.starts_with(&old_pallet_prefix)));
-	}
+	let old_pallet_prefix_iter = frame_support::storage::KeyPrefixIterator::new(
+		old_pallet_prefix.to_vec(),
+		old_pallet_prefix.to_vec(),
+		|_| Ok(()),
+	);
+	assert_eq!(old_pallet_prefix_iter.count(), 0);
```
Comment (thiolliere) [inline frame/membership/src/migrations/v4.rs:122]: is the past  we only checked for the first key at the old prefix, and we accepted that it was the storage_version_key, now we ensure nothing remains there.

File: frame/membership/src/migrations/v4.rs [hunk 3]
```diff
@@ -119,26 +112,24 @@ pub fn post_migrate<P: GetStorageVersion, N: AsRef<str>>(old_pallet_name: N, new
...
 	let new_pallet_prefix = twox_128(new_pallet_name.as_bytes());
-	// Assert that the storages have been moved to the new prefix
-	assert!(storage::next_key(&new_pallet_prefix)
-		.map_or(true, |next_key| next_key.starts_with(&new_pallet_prefix)));
+	let new_pallet_prefix_iter = frame_support::storage::KeyPrefixIterator::new(
+		new_pallet_prefix.to_vec(),
+		new_pallet_prefix.to_vec(),
+		|_| Ok(()),
+	);
+	assert!(new_pallet_prefix_iter.count() > 1);
```
Comment (thiolliere) [inline frame/membership/src/migrations/v4.rs:132]: in the past we ensure a key was at the new prefix, considering the storage_version_key is expected there, it wasn't checking much. Now we ensure at least another one key is there.
Actually maybe membership is in a state when nothing is stored. in this case the post_migration gives a false positive

**Review by shawntabrizi — APPROVED** [review]

File: frame/membership/src/migrations/v4.rs (hunk 1 above)
Comment (emostov) [reply frame/membership/src/migrations/v4.rs:101]: tiny nit, but this doesn't check length, right? Would it be more accurate to do:

```
assert!(new_pallet_prefix_iter.next(), Some(storage_version_key));
assert!(new_pallet_prefix_iter.next(), None);
```
?

**Review by emostov — APPROVED** [review]

File: frame/membership/src/migrations/v4.rs (hunk 1 above)
Comment (thiolliere) [reply frame/membership/src/migrations/v4.rs:101]: yes it is true, but I was thinking if they change the name, maybe it is ok not to have anything stored at the new prefix.

File: frame/membership/src/migrations/v4.rs [hunk 4]
```diff
@@ -85,28 +84,22 @@ pub fn pre_migrate<P: GetStorageVersion, N: AsRef<str>>(old_pallet_name: N, new_
...
-		HexDisplay::from(&storage::next_key(&new_pallet_prefix).unwrap()),
+	let storage_version_key = twox_128(STORAGE_VERSION_STORAGE_KEY_POSTFIX);
+
+	let mut new_pallet_prefix_iter = frame_support::storage::KeyPrefixIterator::new(
+		new_pallet_prefix.to_vec(),
+		new_pallet_prefix.to_vec(),
+		|key| Ok(key.to_vec()),
 	);
+
+	// Ensure nothing except the storage_version_key is stored in the new prefix.
```
Comment (thiolliere) [inline frame/membership/src/migrations/v4.rs:100]:
```suggestion
	// Ensure nothing except maybe the storage_version_key is stored in the new prefix.
```

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**